frame/support/src/traits/hooks.rs

```diff
@@ -239,64 +190,137 @@ impl OnRuntimeUpgrade for Tuple {
-/// Type that provide some integrity tests.
-///
-/// This implemented for modules by `decl_module`.
+/// See [`Hooks::integrity_test`].
 #[cfg_attr(all(not(feature = "tuples-96"), not(feature = "tuples-128")), impl_for_tuples(64))]
 #[cfg_attr(all(feature = "tuples-96", not(feature = "tuples-128")), impl_for_tuples(96))]
 #[cfg_attr(feature = "tuples-128", impl_for_tuples(128))]
 pub trait IntegrityTest {
-	/// Run integrity test.
-	///
-	/// The test is not executed in a externalities provided environment.
+	/// See [`Hooks::integrity_test`].
 	fn integrity_test() {}
 }
 
-/// The pallet hooks trait. Implementing this lets you express some logic to execute.
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// The pallet hooks trait. This is merely an umbrella trait for:
+///
+/// - [`OnInitialize`]
+/// - [`OnFinalize`]
+/// - [`OnRuntimeUpgrade`]
+/// - [`OffchainWorker`]
+/// - [`OnIdle`]
+/// - [`IntegrityTest`]
+///
+/// ## Ordering
+///
+/// For all hooks, except [`OnIdle`] the order of execution is derived from how the pallets are
+/// ordered in [`crate::construct_runtime`].
+///
+/// ## Summary
+///
+/// In short, the following diagram shows the flow of hooks in a pallet
+///
+/// ```mermaid
+/// graph LR
+/// 	Optional --> BeforeExtrinsics
+/// 	BeforeExtrinsics --> Extrinsics
+/// 	Extrinsics --> AfterExtrinsics
+/// 	subgraph Optional
+/// 	OnRuntimeUpgrade
+/// end
+///
+/// subgraph BeforeExtrinsics
+/// 	OnInitialize
+/// end
+///
+/// subgraph Extrinsics
+/// 	direction TB
+/// 	Inherent1
+/// 	Inherent2
+/// 	Extrinsic1
+/// 	Extrinsic2
+///
+/// 	Inherent1 --> Inherent2
+/// 	Inherent2 --> Extrinsic1
+/// 	Extrinsic1 --> Extrinsic2
+/// end
+///
+/// subgraph AfterExtrinsics
+/// 	OnIdle
+/// 	OnFinalize
+///
+/// 	OnIdle --> OnFinalize
+/// end
+/// ```
 pub trait Hooks<BlockNumber> {
-	/// The block is being finalized. Implement to have something happen.
+	/// Block initialization hook. This is called at the very beginning of block execution.
+	///
+	/// Must return the non-negotiable weight of both itself and whatever [`Hooks::on_finalize`]
+	/// wishes to consume.
+	///
+	/// The weight returned by this is treated as `DispatchClass::Mandatory`, meaning that
+	/// it MUST BE EXECUTED. This implies that this hook should only ever be used if the logic in it
+	/// must happen under any circumstance.
+	///
+	/// NOTE: This function is called BEFORE ANY extrinsic in a block is applied, including inherent
+	/// extrinsics. Hence for instance, if you runtime includes `pallet-timestamp`, the `timestamp`
+	/// is not yet up to date at this point.
+	fn on_initialize(_n: BlockNumber) -> Weight {
+		Weight::zero()
+	}
+
+	/// Block finalization hook. This is called at the very end of block execution.
+	///
+	/// Note that this has nothing to do with finality in the "consensus" sense.
+	///
+	/// Note that the non-negotiable weight for this has must have already been returned by
+	/// [`Hooks::on_initialize`]. It usage along is not permitted.
```

Comment:
> It usage along is not permitted.


I don't get what this means.

Reply:
usage "_alone_" 🙈 I meant to say you cannot use this without `on_initialize`.